File: core/x/javaapi/api/src/main/java/viaduct/java/api/context/ConnectionFieldExecutionContext.java
package viaduct.java.api.context;

import viaduct.java.api.types.Connection;
import viaduct.java.api.types.ConnectionArguments;
import viaduct.java.api.types.GraphQLObject;
import viaduct.java.api.types.Query;

/**
 * A {@link FieldExecutionContext} provided to resolvers of connection fields.
 *
 * <p>Narrows the arguments type to {@link ConnectionArguments} and the output type to {@link
 * Connection}, so a connection resolver's {@code ctx.getArguments()} is typed with the pagination
 * inputs. Java equivalent of Kotlin's {@code viaduct.api.context.ConnectionFieldExecutionContext}.
 *
 * @param <T> the type of the object on which this field is being resolved
 * @param <Q> the type of the Query root object
 * @param <A> the connection-arguments type for this field
 * @param <R> the connection output type
 */
public interface ConnectionFieldExecutionContext<
        T extends GraphQLObject,
        Q extends Query,
        A extends ConnectionArguments,
        R extends Connection<?, ?>>
    extends FieldExecutionContext<T, Q, A, R> {}

File: core/tenant/api/src/main/kotlin/viaduct/api/context/ConnectionFieldExecutionContext.kt
package viaduct.api.context

import viaduct.api.types.Connection
import viaduct.api.types.ConnectionArguments
import viaduct.api.types.Object
import viaduct.api.types.Query
import viaduct.apiannotations.ExperimentalApi

/**
 * A [FieldExecutionContext] for connection fields, with typed access to [ConnectionArguments].
 *
 * Resolvers typically don't interact with the arguments directly — [viaduct.api.internal.ConnectionBuilder]
 * reads them internally when building edges and `pageInfo`. Call `ctx.arguments.toOffsetLimit()`
 * directly only when you need the offset/limit to pass to a backend before using the builder.
 *
 * @param O The parent object type containing this connection field.
 * @param Q The query type for data fetching.
 * @param A The connection arguments type (`first`, `after`, `last`, `before`).
 * @param R The connection output type.
 * @see FieldExecutionContext
 * @see viaduct.api.internal.ConnectionBuilder
 */

@ExperimentalApi
interface ConnectionFieldExecutionContext<
    O : Object,
    Q : Query,
    A : ConnectionArguments,
    R : Connection<*, *>,
> : FieldExecutionContext<O, Q, A, R>
